readthedocs/core/mixins.py

```diff
@@ -17,3 +18,47 @@ def get_context_data(self, **kwargs):
 class PrivateViewMixin(LoginRequiredMixin):
 
     pass
+
+
+class UpdateChangeReasonForm:
+
+    """Set the change_reason on the model changed through this form."""
+
+    change_reason = 'Changed from: form'
+
+    def save(self, commit=True):
+        obj = super().save(commit=commit)
+        update_change_reason(obj, self.change_reason)
+        return obj
+
+
+class UpdateChangeReasonAdmin:
+
+    """Set the change_reason on the model changed through this admin view."""
+
+    change_reason = 'Changed from: admin'
+
+    def save_model(self, request, obj, form, change):
+        super().save_model(request, obj, form, change)
+        update_change_reason(obj, self.change_reason)
+
+    def delete_model(self, request, obj):
+        super().delete_model(request, obj)
+        update_change_reason(obj, self.change_reason)
+
+
+class UpdateChangeReasonPostView:
+
+    """
+    Set the change_reason on the model changed through the POST method of this view.
+
+    Use this class for views that don't use a form, like ``DeleteView``.
+    """
+
+    change_reason = 'Changed from: form'
```

Comment:
This doesn't seem super useful -- can we not provide more context here? I guess we'll see how this works in practice, but I feel like it would be more useful to know _what_ form or view? Not worth a ton of extra work though, if it isn't easy.

Reply:
It can be changed, but not sure if provides more context, we only have one form/view per model.

Reply:
I also think we could save this data as json (string in the bd), so is easy to add more metadate, like

```json
{"origin": "form", "class": "ProjectForm"}
```

Reply:
I have changed the message to include the class name and have a more structured form, like we have in logs


![Screenshot 2021-07-22 at 18-21-26 Change history james bond (admin) Django site admin](https://user-images.githubusercontent.com/4975310/126720750-18ebe726-a7b3-4c65-a5d2-272f7ccbc23f.png)

Reply:
Maybe adding the URL's path here is useful as well.

Reply:
could also be an extra field to populate it from the signal since we have access to the request there.